crates/cosmos/cosmos-chain-components/src/impls/transaction/query_tx_response.rs

```diff
@@ -39,6 +47,41 @@ where
             .await
             .map_err(Chain::raise_error)?;
 
-        Ok(response.txs.into_iter().next())
+        let m_response = responses.txs.into_iter().next();
+
+        match m_response {
+            Some(response) => {
+                // Try to ensure that the chain progress to a later height than the height that
+                // returns the response, so that side effects such as merkle proofs can always
+                // be queried from the chain thereafter.
+                //
+                // This is usually not an issue in local tests. But when interacting with public
+                // RPC servers, at the first moment the transaction becomes available, it is not
+                // always propogated to all full nodes yet. This would result in the relayer
+                // interacting with a past chain state that did not have the transaction available
+                // yet, resulting in errors such as ack not found.
+                //
+                // The best way to work around this is to have the relayer always interact with only
+                // one dedicated full node. That way, when the full node says that the transaction
+                // is already there, then it is guaranteed to always available there onward.
+                // But we need this workaround so that the relayer still works with multiple full
+                // nodes and with public RPC endpoints.
+                for _ in 0..20 {
+                    let response_height: u64 = response.height.into();
+                    let chain_height = chain.query_chain_height().await?;
+
+                    // We can technically wait for the same height with >=, but we use > to be
+                    // a little safer to ensure that all full nodes have processed the transaction.
+                    if Chain::revision_height(&chain_height) > response_height {
+                        break;
+                    } else {
+                        chain.runtime().sleep(Duration::from_millis(500)).await;
```

Comment:
Should this be the polling interval of the chain types?

Reply:
Good point. Updated in 2589b27dab2f4e0d7285e022e504d0d8a40844ae.